### ISSUE TITLE
[BUG] tap cannot find dependent module

### ISSUE THREAD

**Issue by pawel-id** [tapjs/tapjs#991]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Have you read the `CONTRIBUTING` guide on posting bugs, and `CODE_OF_CONDUCT`?

- [X] yes I read the things

### This issue exists in the latest tap version

- [X] I am using the latest tap

### Description

I have a problem with running `ts-node` on typescript project (Cannot find module...). This problem happening only while using `ts-node`. This issue is not happening when transpiling typescript with `tsc` and then running javascript directly. It is not happening also while using `tsx` or `tsimp`. 

I spotted that the issue is happening also while running `tap`.

```
npm t

> ts-sandbox@1.0.0 test
> tap


> @tapjs/test-built@0.0.0 prepare
> tshy

 SKIP  test/hello.ts 0 OK 521ms
 ~ no tests found

2> test/hello.ts
node:internal/modules/cjs/loader:1144
  const err = new Error(message);
              ^

Error: Cannot find module '../src/hello.js'
Require stack:
- /Users/pawel/wrk/github/ts-sandbox/test/hello.ts
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at Module._load (node:internal/modules/cjs/loader:985:27)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (/Users/pawel/wrk/github/ts-sandbox/test/hello.ts:2:1)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._compile (/Users/pawel/wrk/github/ts-sandbox/node_modules/pirates/lib/index.js:117:24)
    at Module.m._compile (/Users/pawel/wrk/github/ts-sandbox/node_modules/@tapjs/test/node_modules/@isaacs/ts-node-temp-fork-for-pr-2009/src/index.ts:1368:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at require.extensions.<computed> (/Users/pawel/wrk/github/ts-sandbox/node_modules/@tapjs/test/node_modules/@isaacs/ts-node-temp-fork-for-pr-2009/src/index.ts:1371:12)
 {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/Users/pawel/wrk/github/ts-sandbox/test/hello.ts' ]
}

Node.js v20.10.0
```

I was thinking that the issue is somehow related to `ts-node`. Then following [advice](https://github.com/tapjs/tapjs/issues/976#issuecomment-1824784507) I tried to replace `ts-node` with `tsx` with the following commands:

```
npx tap plugin add @tapjs/tsx
npx tap plugin rm @tapjs/typescript 
```

this operation did a change in package.json adding the following content.

```
{
  "devDependencies": {
    "@tapjs/tsx": "^1.1.18",
  },
  "tap": {
    "plugin": [
      "!@tapjs/typescript",
      "@tapjs/tsx"
    ]
  }
}

```
This fixed an issue for a while. However when cloning the repository and building and  running test it fails again the same way. 

### Reproduction

I created simple repository to replicate the issue https://github.com/pawel-id/ts-sandbox . In order to replicate the issue it should be cloned and the commands should be run:

```shell
$ npm install
$ npm run build
$ npm test
# this will fail with `Cannot find module...`
$ npm run build
$ npm test
# this will pass, but why?
```

Here is [action script](https://github.com/pawel-id/ts-sandbox/blob/main/.github/workflows/main.yml#L29-L33) with [execution log](https://github.com/pawel-id/ts-sandbox/actions/runs/7363574396/job/20043157463).

I wonder why it fails for the first time, and why it pass during the second run.

Note: I do edit of reproduction since initial issue creation.

### Environment

```
which tap
npm ls tap
tap versions
tap config list
tap plugin list
# if using typescript:
npm ls @isaacs/ts-node-temp-fork-for-pr-2009
npm ls typescript
# posix:
uname -a
# windows (in powershell)
"$([Environment]::OSVersion.VersionString) $(('x86', 'x64')[[Environment]::Is64BitOperatingSystem])"
```

```
~/wrk/github/ts-sandbox (main) $ npm ls tap
ts-sandbox@1.0.0 /Users/pawel/wrk/github/ts-sandbox
└── tap@18.6.1

~/wrk/github/ts-sandbox (main) $ npx tap versions
tap: 18.6.1
"@tapjs/config": 2.4.14
"@tapjs/core": 1.4.6
"@tapjs/run": 1.4.16
"@tapjs/stack": 1.2.7
"@tapjs/test": 1.3.17
tap-parser: 15.3.1
tap-yaml: 2.2.1
tcompare: 6.4.5
plugins:
  "@tapjs/after": 1.1.17
  "@tapjs/after-each": 1.1.17
  "@tapjs/asserts": 1.1.17
  "@tapjs/before": 1.1.17
  "@tapjs/before-each": 1.1.17
  "@tapjs/filter": 1.2.17
  "@tapjs/fixture": 1.2.17
  "@tapjs/intercept": 1.2.17
  "@tapjs/mock": 1.2.15
  "@tapjs/node-serialize": 1.2.6
  "@tapjs/snapshot": 1.2.17
  "@tapjs/spawn": 1.1.17
  "@tapjs/stdin": 1.1.17
  "@tapjs/tsx": 1.1.18
  "@tapjs/worker": 1.1.17
~/wrk/github/ts-sandbox (main) $ npx tap config list
# vim: set filetype=yaml :

# from package.json
plugin:
  - "!@tapjs/typescript"
  - "@tapjs/tsx"

# env, cli, and defaults
color: true
coverage-report:
  - text
exclude:
  - "**/@(fixture*(s)|dist)/**"
include:
  - "**/@(test?(s)|__test?(s)__)/**/*.@(js|cjs|mjs|tap|cts|jsx|mts|ts|tsx)"
  - "**/*.@(test?(s)|spec).@(js|cjs|mjs|tap|cts|jsx|mts|ts|tsx)"
  - "**/test?(s).@(js|cjs|mjs|tap|cts|jsx|mts|ts|tsx)"
jobs: 6
reporter: base
snapshot-clean-cwd: true
timeout: 30

~/wrk/github/ts-sandbox (main) $ npx tap plugin list
@tapjs/after
@tapjs/after-each
@tapjs/asserts
@tapjs/before
@tapjs/before-each
@tapjs/filter
@tapjs/fixture
@tapjs/intercept
@tapjs/mock
@tapjs/node-serialize
@tapjs/snapshot
@tapjs/spawn
@tapjs/stdin
@tapjs/worker
@tapjs/tsx

~/wrk/github/ts-sandbox (main) $ npm ls @isaacs/ts-node-temp-fork-for-pr-2009
ts-sandbox@1.0.0 /Users/pawel/wrk/github/ts-sandbox
└─┬ tap@18.6.1
  ├─┬ @tapjs/test@1.3.17
  │ └── @isaacs/ts-node-temp-fork-for-pr-2009@10.9.5
  └─┬ @tapjs/typescript@1.3.6
    └── @isaacs/ts-node-temp-fork-for-pr-2009@10.9.5

~/wrk/github/ts-sandbox (main) $ npm ls typescript
ts-sandbox@1.0.0 /Users/pawel/wrk/github/ts-sandbox
├─┬ tap@18.6.1
│ ├─┬ @tapjs/test@1.3.17
│ │ ├─┬ @isaacs/ts-node-temp-fork-for-pr-2009@10.9.5
│ │ │ └── typescript@5.2.2 deduped
│ │ ├─┬ tshy@1.8.2
│ │ │ └── typescript@5.2.2
│ │ └── typescript@5.2.2
│ └─┬ @tapjs/typescript@1.3.6
│   └─┬ @isaacs/ts-node-temp-fork-for-pr-2009@10.9.5
│     └── typescript@5.3.3 deduped
├─┬ ts-node@10.9.2
│ └── typescript@5.3.3 deduped
├─┬ tsimp@2.0.10
│ └── typescript@5.3.3 deduped
└── typescript@5.3.3

~/wrk/github/ts-sandbox (main) $ uname -a
Darwin pixel.local 23.2.0 Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000 arm64
```

**Comment by pawel-id** [comment]
I edited a reproduction steps and added link to github action log showing the issue.

**Comment by pawel-id** [comment]
With a little help of stackoverflow I found another workaround to `ts-node` issue and also as a side effect fixing this issue with `tap`.  I removed .js suffixes from imports and it fixed all problems. I.e. the `import hello from '../src/hello.js'` was replaced by `import hello from '../src/hello'`.

I feel that the problem lies somewhere in 'ts-node' which has some issues with the import having .js suffix. `tsc` compiler has no issues with both forms (with .js and without).

Anyway I will leave this open, because the mystery of `tap` test being fixed on second run probably require some investigation.
